File: Packages/StrandAnalytics/Sources/StrandAnalytics/SleepStress.swift
import Foundation
import WhoopProtocol

/// Evidence-gated overnight autonomic load on the same 0...3 scale as DaytimeStress.
///
/// Every published point needs both dense heart rate and a clean R-R-derived RMSSD value.
/// The whole result fails closed unless at least 60% of the sleep window is covered. This is
/// a NOOP estimate, not WHOOP's proprietary Sleep Stress score or a clinical measurement.
public enum SleepStress {
    public static let bucketSeconds = 300
    public static let minHRSamplesPerBucket = 60
    public static let minCoveredBuckets = 6
    public static let minCoverageFraction = 0.60
    public static let mediumBandFloor = 1.0
    public static let highBandFloor = 2.0

    /// A sleep-specific offset keeps an ordinary resting window in the low band. Excursions
    /// in HR and/or drops in RMSSD must rise materially above the night's calm reference.
    static let sleepBaselineOffset = 2.0

    public enum Band: String, Equatable, Sendable {
        case low
        case medium
        case high
    }

    public struct Point: Equatable, Sendable {
        public let startTs: Int
        public let level: Double
        public let meanHR: Double
        public let rmssd: Double

        public var band: Band {
            if level >= SleepStress.highBandFloor { return .high }
            if level >= SleepStress.mediumBandFloor { return .medium }
            return .low
        }

        public init(startTs: Int, level: Double, meanHR: Double, rmssd: Double) {
            self.startTs = startTs
            self.level = level
            self.meanHR = meanHR
            self.rmssd = rmssd
        }
    }

    public struct Result: Equatable, Sendable {
        public let points: [Point]
        public let eligibleBucketCount: Int
        public let coverageFraction: Double

        public func bucketCount(in band: Band) -> Int {
            points.lazy.filter { $0.band == band }.count
        }

        public func fraction(in band: Band) -> Double {
            guard !points.isEmpty else { return 0 }
            return Double(bucketCount(in: band)) / Double(points.count)
        }

        public func durationSeconds(in band: Band) -> Int {
            bucketCount(in: band) * SleepStress.bucketSeconds
        }

        public init(points: [Point], eligibleBucketCount: Int, coverageFraction: Double) {
            self.points = points
            self.eligibleBucketCount = eligibleBucketCount
            self.coverageFraction = coverageFraction
        }
    }

    /// Returns nil unless the night has enough jointly-covered HR and clean R-R windows.
    public static func analyze(
        hr: [HRSample],
        rr: [RRInterval],
        startTs: Int,
        endTs: Int
    ) -> Result? {
        guard endTs > startTs else { return nil }
        let eligibleCount = (endTs - startTs) / bucketSeconds
        guard eligibleCount >= minCoveredBuckets else { return nil }

        var hrByBucket = Array(repeating: [Double](), count: eligibleCount)
        var rrByBucket = Array(repeating: [Double](), count: eligibleCount)

        for sample in hr where sample.ts >= startTs && sample.ts < endTs {
            let index = (sample.ts - startTs) / bucketSeconds
            if index >= 0 && index < eligibleCount {
                hrByBucket[index].append(Double(sample.bpm))
            }
        }
        for sample in rr where sample.ts >= startTs && sample.ts < endTs {
            let index = (sample.ts - startTs) / bucketSeconds
            if index >= 0 && index < eligibleCount {
                rrByBucket[index].append(Double(sample.rrMs))
            }
        }

        struct Aggregate {
            let index: Int
            let meanHR: Double
            let rmssd: Double
        }

        var aggregates: [Aggregate] = []
        aggregates.reserveCapacity(eligibleCount)
        for index in 0..<eligibleCount {
            let hrs = hrByBucket[index]
            guard hrs.count >= minHRSamplesPerBucket else { continue }
            let hrv = HRVAnalyzer.analyze(rawRR: rrByBucket[index])
            guard hrv.nClean >= HRVAnalyzer.minBeats,
                  let rmssd = hrv.rmssd, rmssd.isFinite else { continue }
            let meanHR = hrs.reduce(0, +) / Double(hrs.count)
            guard meanHR.isFinite else { continue }
            aggregates.append(Aggregate(index: index, meanHR: meanHR, rmssd: rmssd))
        }

        let coverage = Double(aggregates.count) / Double(eligibleCount)
        guard aggregates.count >= minCoveredBuckets,
              coverage >= minCoverageFraction else { return nil }

        let heartRates = aggregates.map(\.meanHR)
        let rmssdValues = aggregates.map(\.rmssd)
        let calmHR = quantile(heartRates.sorted(), q: 0.25)
        let calmRMSSD = quantile(rmssdValues.sorted(), q: 0.75)
        let sdHR = standardDeviation(heartRates)
        let sdRMSSD = standardDeviation(rmssdValues)

        let points = aggregates.map { aggregate in
            var raw = 0.0
            if sdHR > 0.0001 {
                raw += (aggregate.meanHR - calmHR) / sdHR
            }
            if sdRMSSD > 0.0001 {
                raw += (calmRMSSD - aggregate.rmssd) / sdRMSSD
            }
            let level = 3.0 / (1.0 + exp(-(raw - sleepBaselineOffset)))
            return Point(
                startTs: startTs + aggregate.index * bucketSeconds,
                level: min(max(level, 0), 3),
                meanHR: aggregate.meanHR,
                rmssd: aggregate.rmssd)
        }

        return Result(
            points: points,
            eligibleBucketCount: eligibleCount,
            coverageFraction: coverage)
    }

    private static func standardDeviation(_ values: [Double]) -> Double {
        guard values.count > 1 else { return 0 }
        let mean = values.reduce(0, +) / Double(values.count)
        let variance = values.reduce(0) { $0 + ($1 - mean) * ($1 - mean) }
            / Double(values.count)
        return variance.squareRoot()
    }

    private static func quantile(_ sorted: [Double], q: Double) -> Double {
        guard let first = sorted.first else { return 0 }
        guard sorted.count > 1 else { return first }
        let position = min(max(q, 0), 1) * Double(sorted.count - 1)
        let lower = Int(position.rounded(.down))
        let upper = min(lower + 1, sorted.count - 1)
        let fraction = position - Double(lower)
        return sorted[lower] + fraction * (sorted[upper] - sorted[lower])
    }
}

File: android/app/src/main/java/com/noop/analytics/SleepStress.kt
package com.noop.analytics

import com.noop.data.HrSample
import com.noop.data.RrInterval
import kotlin.math.exp
import kotlin.math.sqrt

/**
 * Evidence-gated overnight autonomic load on the same 0..3 scale as [DaytimeStress].
 *
 * Every published point needs both dense heart rate and a clean R-R-derived RMSSD value.
 * The whole result fails closed unless at least 60% of the sleep window is covered. This is
 * a NOOP estimate, not WHOOP's proprietary Sleep Stress score or a clinical measurement.
 */
object SleepStress {
    const val BUCKET_SECONDS = 300L
    const val MIN_HR_SAMPLES_PER_BUCKET = 60
    const val MIN_COVERED_BUCKETS = 6
    const val MIN_COVERAGE_FRACTION = 0.60
    const val MEDIUM_BAND_FLOOR = 1.0
    const val HIGH_BAND_FLOOR = 2.0

    /** Keeps an ordinary resting window low until HR rises and/or RMSSD falls materially. */
    private const val SLEEP_BASELINE_OFFSET = 2.0

    enum class Band { LOW, MEDIUM, HIGH }

    data class Point(
        val startTs: Long,
        val level: Double,
        val meanHr: Double,
        val rmssd: Double,
    ) {
        val band: Band
            get() = when {
                level >= HIGH_BAND_FLOOR -> Band.HIGH
                level >= MEDIUM_BAND_FLOOR -> Band.MEDIUM
                else -> Band.LOW
            }
    }

    data class Result(
        val points: List<Point>,
        val eligibleBucketCount: Int,
        val coverageFraction: Double,
    ) {
        fun bucketCount(band: Band): Int = points.count { it.band == band }

        fun fraction(band: Band): Double =
            if (points.isEmpty()) 0.0 else bucketCount(band).toDouble() / points.size

        fun durationSeconds(band: Band): Long = bucketCount(band) * BUCKET_SECONDS
    }

    /** Returns null unless the night has enough jointly-covered HR and clean R-R windows. */
    fun analyze(
        hr: List<HrSample>,
        rr: List<RrInterval>,
        startTs: Long,
        endTs: Long,
    ): Result? {
        if (endTs <= startTs) return null
        val eligibleCount = ((endTs - startTs) / BUCKET_SECONDS).toInt()
        if (eligibleCount < MIN_COVERED_BUCKETS) return null

        val hrByBucket = List(eligibleCount) { ArrayList<Double>() }
        val rrByBucket = List(eligibleCount) { ArrayList<Double>() }

        for (sample in hr) {
            if (sample.ts < startTs || sample.ts >= endTs) continue
            val index = ((sample.ts - startTs) / BUCKET_SECONDS).toInt()
            if (index in 0 until eligibleCount) hrByBucket[index].add(sample.bpm.toDouble())
        }
        for (sample in rr) {
            if (sample.ts < startTs || sample.ts >= endTs) continue
            val index = ((sample.ts - startTs) / BUCKET_SECONDS).toInt()
            if (index in 0 until eligibleCount) rrByBucket[index].add(sample.rrMs.toDouble())
        }

        data class Aggregate(val index: Int, val meanHr: Double, val rmssd: Double)

        val aggregates = ArrayList<Aggregate>(eligibleCount)
        for (index in 0 until eligibleCount) {
            val hrs = hrByBucket[index]
            if (hrs.size < MIN_HR_SAMPLES_PER_BUCKET) continue
            val hrv = HrvAnalyzer.analyzeRaw(rrByBucket[index])
            val rmssd = hrv.rmssd
            if (hrv.nClean < HrvAnalyzer.MIN_BEATS || rmssd == null || !rmssd.isFinite()) continue
            val meanHr = hrs.average()
            if (!meanHr.isFinite()) continue
            aggregates.add(Aggregate(index, meanHr, rmssd))
        }

        val coverage = aggregates.size.toDouble() / eligibleCount
        if (aggregates.size < MIN_COVERED_BUCKETS || coverage < MIN_COVERAGE_FRACTION) return null

        val heartRates = aggregates.map { it.meanHr }
        val rmssdValues = aggregates.map { it.rmssd }
        val calmHr = quantile(heartRates.sorted(), 0.25)
        val calmRmssd = quantile(rmssdValues.sorted(), 0.75)
        val sdHr = standardDeviation(heartRates)
        val sdRmssd = standardDeviation(rmssdValues)

        val points = aggregates.map { aggregate ->
            var raw = 0.0
            if (sdHr > 0.0001) raw += (aggregate.meanHr - calmHr) / sdHr
            if (sdRmssd > 0.0001) raw += (calmRmssd - aggregate.rmssd) / sdRmssd
            val level = (3.0 / (1.0 + exp(-(raw - SLEEP_BASELINE_OFFSET)))).coerceIn(0.0, 3.0)
            Point(
                startTs = startTs + aggregate.index * BUCKET_SECONDS,
                level = level,
                meanHr = aggregate.meanHr,
                rmssd = aggregate.rmssd,
            )
        }

        return Result(points, eligibleCount, coverage)
    }

    private fun standardDeviation(values: List<Double>): Double {
        if (values.size <= 1) return 0.0
        val mean = values.average()
        return sqrt(values.sumOf { (it - mean) * (it - mean) } / values.size)
    }

    private fun quantile(sorted: List<Double>, q: Double): Double {
        if (sorted.isEmpty()) return 0.0
        if (sorted.size == 1) return sorted.first()
        val position = q.coerceIn(0.0, 1.0) * (sorted.size - 1)
        val lower = position.toInt()
        val upper = minOf(lower + 1, sorted.lastIndex)
        val fraction = position - lower
        return sorted[lower] + fraction * (sorted[upper] - sorted[lower])
    }
}
